# Tutorial: learning genotype-to-phenotype models with Kover

This tutorial will walk you through an application of Kover to a set of genomes labelled according to their phenotypes. We use data from the paper (see [data](../data/)), but you could use your own.

## Outline

* [Getting the example data](#getting-the-example-data)
* [Creating a Kover dataset](#creating-a-kover-dataset)
* [Splitting the data into training and testing sets](#splitting-the-data-into-training-and-testing-sets)
* [Learning models with:](#learning-models)
  * [Set Covering Machines](#set-covering-machines)
  * [Classification and Regression Trees](#classification-and-regression-trees)
* [Interpreting the learned models](#interpreting-the-learned-models)
  * [Annotating k-mers](#annotating-k-mers)
  * [Analyzing equivalent rules](#analyzing-equivalent-rules)


## Getting the example data

## Creating a Kover dataset

## Splitting the data into training and testing sets

## Learning models

### Set Covering Machines

### Classification and Regression Trees
** Dont forget to mention multiclass can be done and how **

## Interpreting the learned models

### Annotating k-mers
Show a simple example using nucleotide BLAST

### Analyzing equivalent rules
Show a simple example using UGENE
